discovery.go

```diff
@@ -0,0 +1,235 @@
+package pubsub
+
+import (
+	"context"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/discovery"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+// discover represents the discovery pipeline.
+// The discovery pipeline handles advertising and discovery of peers
+type discover struct {
+	p *PubSub
+
+	// discovery assists in discovering and advertising peers for a topic
+	discovery discovery.Discovery
+
+	// advertising tracks which topics are being advertised
+	advertising map[string]context.CancelFunc
+
+	// discoverQ handles continuing peer discovery
+	discoverQ chan *discoverReq
+
+	// ongoing tracks ongoing discovery requests
+	ongoing map[string]struct{}
+
+	// done handles completion of a discovery request
+	done chan string
+}
+
+type IsDiscovered func(numPeers int) bool
+
+func CheckNumPeers(num int) IsDiscovered {
+	return func(numPeers int) bool {
+		return numPeers >= num
+	}
+}
+
+// newDiscover creates a new peer discovery mechanism
+func newDiscover() *discover {
+	return &discover{
+		advertising: make(map[string]context.CancelFunc),
+		discoverQ:   make(chan *discoverReq, 32),
+		ongoing:     make(map[string]struct{}),
+		done:        make(chan string),
+	}
+}
+
+// Start attaches the discovery pipeline to a pubsub instance and starts event loop
+func (d *discover) Start(p *PubSub) {
+	if d.discovery == nil || p == nil {
+		return
+	}
+
+	d.p = p
+	go d.discoverLoop()
+}
+
+func (d *discover) discoverLoop() {
+	for {
+		select {
+		case discover := <-d.discoverQ:
+			topic := discover.topic
+
+			if _, ok := d.ongoing[topic]; ok {
+				discover.done <- struct{}{}
```

Comment:
Should we really mark this as done? Doesn't that make this channel completely useless?

Reply:
in the Bootstrap function we wait for done before we try again to see if we're bootstrapped.